waspc/test/Analyzer/ParserTest.hs

```diff
@@ -5,319 +5,337 @@ import Data.Either (isLeft)
-            unlines
-              [ "test Decl {",
-                "  string: \"Hello Wasp =}\",",
-                "  escapedString: \"Look, a \\\"\",",
-                "  integer: 42,",
-                "  real: 3.14,",
-                "  yes: true,",
-                "  no: false,",
-                "  ident: Wasp,",
-                "  // This is a comment",
-                "  innerDict: { innerDictReal: 2.17 }",
-                "}"
-              ]
-      let ast =
-            AST
-              [ wctx (1, 1) (11, 1) $
-                  Decl "test" "Decl" $
-                    wctx (1, 11) (11, 1) $
-                      Dict
-                        [ ("string", wctx (2, 11) (2, 25) $ StringLiteral "Hello Wasp =}"),
-                          ("escapedString", wctx (3, 18) (3, 29) $ StringLiteral "Look, a \""),
-                          ("integer", wctx (4, 12) (4, 13) $ IntegerLiteral 42),
-                          ("real", wctx (5, 9) (5, 12) $ DoubleLiteral 3.14),
-                          ("yes", wctx (6, 8) (6, 11) $ BoolLiteral True),
-                          ("no", wctx (7, 7) (7, 11) $ BoolLiteral False),
-                          ("ident", wctx (8, 10) (8, 13) $ Var "Wasp"),
-                          ( "innerDict",
-                            wctx (10, 14) (10, 36) $
-                              Dict
-                                [ ("innerDictReal", wctx (10, 31) (10, 34) $ DoubleLiteral 2.17)
-                                ]
-                          )
-                        ]
-              ]
-      parse source `shouldBe` Right ast
+spec_IsValidWaspIdentifier :: Spec
+spec_IsValidWaspIdentifier = do
+  it "Correctly identifies valid identifiers" $ do
+    all isValidWaspIdentifier ["foo", "foo12", "_12", "_"]
+  it "Correctly identifies invalid identifiers" $ do
+    not (any isValidWaspIdentifier ["true", "12", "12_", "1foo"])
 
-    it "Parses comments" $ do
-      let source =
-            unlines
-              [ "  // This is some // comment",
-                "/* comment",
-                "  span//ning",
-                "multi/*ple lines */",
-                "test /* *hi* */ Decl 42  // One more comment",
-                "// And here is final comment"
-              ]
-      let ast =
-            AST
-              [ wctx (5, 1) (5, 23) $
-                  Decl "test" "Decl" $ wctx (5, 22) (5, 23) $ IntegerLiteral 42
-              ]
-      parse source `shouldBe` Right ast
+spec_ParseExpression :: Spec
+spec_ParseExpression = do
+  it "Parses identifiers" $ do
+    parseExpression "foo" `shouldBe` Right (Var "foo")
+    parseExpression "foo12" `shouldBe` Right (Var "foo12")
+    parseExpression "_" `shouldBe` Right (Var "_")
+    parseExpression "_12" `shouldBe` Right (Var "_12")
+  it "Parses double literals" $ do
+    parseExpression "0.5" `shouldBe` Right (DoubleLiteral 0.5)
+  it "Parses int literals" $ do
+    parseExpression "5" `shouldBe` Right (IntegerLiteral 5)
```

Comment:
This is a nice side effect of exposing an expression parser - we can now test the parser more specifically. When something breaks, we're no longer testing a black box that only knows how to parse full Wasp source files.

Reply:
Yup, this is very cool!!